webpack/assets/javascripts/react_app/common/hooks/API/APIHooks.js

```diff
@@ -15,30 +15,30 @@ import { APIActions } from '../../../redux/API';
  * @return {object} returns an object that contains the response, status, key and 'setUrl' for setting the url dynamically
  */
 
-export const useAPI = (method, url, options) => {
+export const useAPI = (method, url, options = {}) => {
   const dispatch = useDispatch();
-  const keyRef = useRef(options?.key);
+  const { key, ...rest } = options;
```

Comment:
How about something like along the lines of:
```suggestion
  const { params, headers, ...rest } = options;
  // these are values that when the caller changes we want a new request to fire, but if they are the same we don't want to change
  const APIparams = useMemo(() => ({params, headers}), [params, headers]);
  // these are the values that shouldn't change, but we still want to make them available without retriggering on rerender and not require them in the effect deps
  const { current: otherParams } = useRef(rest);
```

Reply:
```js
 const APIparams = useMemo(() => ({params, headers}), [params, headers]);
```
`useMemo` uses a shallow comparison, when the `params` object contains nested objects (or arrays of objects), it won't be memorized well, and an infinite loop might occur. that's why it makes sense to ask the API consumer to pass a memorized object when needed. For more details about this issue, I recommend following [this issue](https://github.com/facebook/react/issues/14476), and these corresponding [suggestions](https://github.com/facebook/react/issues/14476#issuecomment-471199055)
 
instead of memorizations, another approch can be [use-deep-compare-effect](https://github.com/kentcdodds/use-deep-compare-effect)  which is a `useEffect` with deep comparison ability.